src/tests/ke/suite/dpc.c

```diff
@@ -0,0 +1,279 @@
+#include <xboxkrnl/xboxkrnl.h>
+#include <string.h>
+
+#include "util/output.h"
+#include "assertions/defines.h"
+
+// TODO: Add list of _LEVEL suffix support into nxdk
+#define DPC_DISPATCH_LEVEL 2 // DISPATCH_LEVEL
+
+typedef struct _dpc_test {
+    BOOLEAN inserted;
+    BOOLEAN removal;
+    BOOL dpc_called;
+    KDPC dpcObject;
+} dpc_test;
+
+static ULONG getDpcRoutineActive()
+{
+    ULONG getter;
+    __asm__ ("movl %%fs:0x58, %0"
+            : "=r"(getter));
+    return getter;
+}
+
+static void setDpcRoutineActive(ULONG setter)
+{
+    __asm__ ("movl %0, %%fs:0x58"
+            :
+            : "a"(setter));
+}
+
+static KDPC dpcObject;
+static BOOL dpc_called;
+static ULONG dpc_executing;
+static void __stdcall dpc_callback(
+	IN KDPC* Dpc,
+	IN PVOID DeferredContext,
+	IN PVOID SystemArgument1,
+	IN PVOID SystemArgument2
+)
+{
+    // DPC called variable
+    if (SystemArgument1 && (DWORD)SystemArgument1 > 0xFF) {
+        *(BOOL*)SystemArgument1 = TRUE;
+    }
+
+    // Request KeIsExecutingDpc return value
+    if (SystemArgument2 && (DWORD)SystemArgument2 > 0xFF) {
```

Comment:
One possible reason why the value `0xFF` is used here, might be because a test fills a struct with `0xCC` bytes. (If this is the reason, isn't that resulting in `0xCCCCCCCC`, invalidating the assumption?)

Perhaps the above is nonsense (hard to tell, without an explanatory comment, which is kind of the whole point of these remarks), an alternative explanation  might be that this `0xFF` is used to avoid accessing the zero page (the first 4 kilobytes of memory), which would otherwise trigger part faults. (If this is the reason, it might be better to use `PAGE_SIZE`instead). But this too might not the intent...

Reply:
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L96
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L101-L102

---

https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L111
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L115-L116

---

https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L128
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L160-L161
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L166-L167

The above points out that KeInsertQueueDpc function has SystemArgument# will be set in the structure when inserted. It is half mainly to check if dpc's object structure does get set properly or not. Or the structure provided by nxdk could be wrong, at least it is correct.

It has nothing to do with test_KeInitializeDpc function that sets 0xCCs since it doesn't insert into dpc queue line.

Reply:
Thanks for all the copy-pastes, I've read all that quite a few times before I commented. It's obvious the tests check passed-in arguments are indeed stored where expected. That's good.

Still, the check against 0xFF in `dpc_callback` congres across as arbitrarily, until all tests are investigated. For instance the second argument is once passed in as an actual memory address, to test that `KeInsertQueueDpc` indeed triggers the dpc. Other times it's just used as a recognizable argument value that should not perform that "is the dpc triggered" check.
This all boils down to:
* the suggestion to explain in a comment why the 0xFF is used like this,
* perhaps replacing all magic values with named symbols, 
* Perhaps use `PAGE_SIZE` instead of 0xFF (or create and use a tooling function like `IsValidPointer(PVOID)`)
* Reducing the two comparisons into one

Reply:
I can add additional comment except I'm not planning to replace any of magic values with named symbols at all. Since I preferred to do manual increment for each test performed and the tests are really that small. I can see the point for reduce two comparisons into one. Originally it came as `if (SystemArgument2) {`, then noticed they are stored into the structure. Which had me add `SystemArgument2 > 0xFF` comparison check. I don't plan on use PAGE_SIZE since I am only performing the test within one byte size. Which has nothing to do with memory nor relevant to paging size. As for `IsValidPointer(PVOID)`, I preferred to avoid any kernel APIs, nor exceptions, contacts or otherwise will make the test invalid unintentionally. I'm pretty sure we can't have any unhandled exceptions in dpc thread anyway.

Such example of using nxdk's print to screen and/or write to log file can unintentionally break the test results. Why is that? There are couple manual hacks in place to force mimic how kernel APIs would work.

TLDR edition:
I'll add a comment and reduce comparison to one for both SystemArgument# inputs.

Reply:
If happy with the review remarks in #80 pull request. Please mark this conversation as resolved and approve the changes there.